chan.go

```diff
@@ -128,92 +157,103 @@ func (r *chanResponse) RawNext() (interface{}, error) {
 
 		return nil, io.EOF
 	case <-ctx.Done():
-		close(r.done)
 		return nil, ctx.Err()
 	}
 
 }
 
-type chanResponseEmitter struct {
-	ch   chan<- interface{}
-	wait chan struct{}
-	done <-chan struct{}
-
-	length *uint64
-	err    **cmdkit.Error
-
-	emitted bool
-}
+type chanResponseEmitter chanResponse
 
 func (re *chanResponseEmitter) Emit(v interface{}) error {
+	// Initially this library allowed commands to return errors by sending an
+	// error value along a stream. We removed that in favour of CloseWithError,
+	// so we want to make sure we catch situations where some code still uses the
+	// old error emitting semantics.
+	// Also errors may occur both as pointers and as plain values, so we need to
+	// check both.
+	if e, ok := v.(cmdkit.Error); ok {
+		v = &e
+	}
+	if e, ok := v.(*cmdkit.Error); ok {
+		log.Errorf("unexpected error value emitted: %s at\n%s", e.Error(), debug.Stack())
+	}
+
+	// channel emission iteration
+	// TODO maybe remove this and use EmitChan instead of calling Emit directly?
 	if ch, ok := v.(chan interface{}); ok {
 		v = (<-chan interface{})(ch)
 	}
-
 	if ch, isChan := v.(<-chan interface{}); isChan {
-		for v = range ch {
-			err := re.Emit(v)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
+		return EmitChan(re, ch)
 	}
 
-	re.emitted = true
-	if re.wait != nil {
+	// unblock Length(), Error() and Head()
+	re.waitOnce.Do(func() {
 		close(re.wait)
-		re.wait = nil
-	}
+	})
+
+	re.l.Lock()
+	defer re.l.Unlock()
 
 	if re.ch == nil {
```

Comment:
I don't think this is a valid "closed" check anymore. Probably want to check:

```go
select {
  case <-re.closed:
    return ...
  default:
}
```

Reply:
You're right it's not a valid close check anymore, but I think your proposal is also racy: It might occur that re.closed is not yet closed but re.ch is, which would result in a write to closed channel error. Thinking this through...

Reply:
It should be if we only close the channel while holding the lock.

This is actually a good case for removing the `closeOnce` `sync.Once` (a user may think that `closeOnce` is protecting the close while really the lock is).